From 902f76a6861eb0855dedf6578c4f57e4ee92d1d2 Mon Sep 17 00:00:00 2001
From: Conor McCarthy <cmccarthy@codeweavers.com>
Date: Fri, 3 Jul 2026 13:22:57 +1000
Subject: [PATCH 42/48] wintypes: Implement
 IDataWriterFactory::CreateDataWriter().

(cherry picked from commit d73fd47973bf3cf53233c0a62a3095f3bbae2986)

CW-Bug-Id: #27221
---
 dlls/wintypes/storage.c        | 46 ++++++++++++++++++++++++++++------
 dlls/wintypes/tests/wintypes.c | 11 +++-----
 2 files changed, 42 insertions(+), 15 deletions(-)

diff --git a/dlls/wintypes/storage.c b/dlls/wintypes/storage.c
index 0afc8ade558..b1e7cf93287 100644
--- a/dlls/wintypes/storage.c
+++ b/dlls/wintypes/storage.c
@@ -39,6 +39,7 @@ struct data_writer
     IDataWriter IDataWriter_iface;
     LONG ref;
 
+    IOutputStream *stream;
     IBuffer *buffer;
     byte *data;
 };
@@ -128,6 +129,22 @@ static ULONG WINAPI data_writer_Release(IDataWriter *iface)
 
     if (!ref)
     {
+        if (impl->stream)
+        {
+            IClosable *closable;
+            HRESULT hr;
+
+            if (SUCCEEDED(hr = IOutputStream_QueryInterface(impl->stream, &IID_IClosable, (void **)&closable)))
+            {
+                hr = IClosable_Close(closable);
+                IClosable_Release(closable);
+            }
+
+            if (FAILED(hr))
+                WARN("Failed to close stream, hr %#lx.\n", hr);
+
+            IOutputStream_Release(impl->stream);
+        }
         if (impl->buffer)
             IBuffer_Release(impl->buffer);
         free(impl);
@@ -306,18 +323,30 @@ static HRESULT WINAPI data_writer_MeasureString(IDataWriter *iface, HSTRING valu
 
 static HRESULT WINAPI data_writer_StoreAsync(IDataWriter *iface, IAsyncOperation_UINT32 **operation)
 {
+    struct data_writer *impl = impl_from_IDataWriter(iface);
+
     FIXME("iface %p, operation %p stub!\n", iface, operation);
 
     *operation = NULL;
-    return HRESULT_FROM_WIN32(ERROR_INVALID_OPERATION);
+
+    if (!impl->stream)
+        return HRESULT_FROM_WIN32(ERROR_INVALID_OPERATION);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI data_writer_FlushAsync(IDataWriter *iface, IAsyncOperation_boolean **operation)
 {
+    struct data_writer *impl = impl_from_IDataWriter(iface);
+
     FIXME("iface %p, operation %p stub!\n", iface, operation);
 
     *operation = NULL;
-    return HRESULT_FROM_WIN32(ERROR_INVALID_OPERATION);
+
+    if (!impl->stream)
+        return HRESULT_FROM_WIN32(ERROR_INVALID_OPERATION);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI data_writer_DetachBuffer(IDataWriter *iface, IBuffer **buffer)
@@ -379,7 +408,7 @@ static const struct IDataWriterVtbl data_writer_vtbl =
     data_writer_DetachStream,
 };
 
-static HRESULT data_writer_create(IDataWriter **out)
+static HRESULT data_writer_create(IOutputStream *output_stream, IDataWriter **out)
 {
     struct data_writer *impl;
     HRESULT hr;
@@ -389,6 +418,7 @@ static HRESULT data_writer_create(IDataWriter **out)
         return E_OUTOFMEMORY;
 
     impl->IDataWriter_iface.lpVtbl = &data_writer_vtbl;
+    impl->stream = output_stream;
     impl->ref = 1;
 
     if (FAILED(hr = data_writer_init_buffer(impl, 0)))
@@ -397,6 +427,9 @@ static HRESULT data_writer_create(IDataWriter **out)
         return hr;
     }
 
+    if (output_stream)
+        IOutputStream_AddRef(output_stream);
+
     *out = &impl->IDataWriter_iface;
     return S_OK;
 }
@@ -491,7 +524,7 @@ static HRESULT STDMETHODCALLTYPE data_writer_activation_factory_ActivateInstance
 
     TRACE("iface %p, instance %p.\n", iface, instance);
 
-    if (FAILED(hr = data_writer_create(&data_writer)))
+    if (FAILED(hr = data_writer_create(NULL, &data_writer)))
         return hr;
 
     hr = IDataWriter_QueryInterface(data_writer, &IID_IInspectable, (void **)instance);
@@ -580,9 +613,8 @@ static HRESULT STDMETHODCALLTYPE data_writer_factory_GetTrustLevel(IDataWriterFa
 static HRESULT STDMETHODCALLTYPE data_writer_factory_CreateDataWriter(IDataWriterFactory *iface,
         IOutputStream *output_stream, IDataWriter **data_writer)
 {
-    FIXME("iface %p, output_stream %p, data_writer %p stub!\n", iface, output_stream, data_writer);
-    *data_writer = NULL;
-    return E_NOTIMPL;
+    TRACE("iface %p, output_stream %p, data_writer %p.\n", iface, output_stream, data_writer);
+    return data_writer_create(output_stream, data_writer);
 }
 
 static const struct IDataWriterFactoryVtbl data_writer_factory_vtbl =
diff --git a/dlls/wintypes/tests/wintypes.c b/dlls/wintypes/tests/wintypes.c
index b50229c7b17..cea53a3c62b 100644
--- a/dlls/wintypes/tests/wintypes.c
+++ b/dlls/wintypes/tests/wintypes.c
@@ -1729,14 +1729,8 @@ static void test_DataWriter(void)
     hr = IRandomAccessStream_QueryInterface(in_memory_stream, &IID_IOutputStream, (void **)&output_stream);
     ok(hr == S_OK, "got hr %#lx.\n", hr);
     hr = IDataWriterFactory_CreateDataWriter(data_writer_factory, output_stream, &data_writer);
-    todo_wine
     ok(hr == S_OK, "got hr %#lx.\n", hr);
     IOutputStream_Release(output_stream);
-    if (FAILED(hr))
-    {
-        IActivationFactory_Release(factory);
-        goto done;
-    }
 
     for (i = 0; i < 2; ++i)
     {
@@ -1761,6 +1755,8 @@ static void test_DataWriter(void)
     hr = IDataWriter_StoreAsync(data_writer, &operation);
     todo_wine
     ok(hr == S_OK, "got hr %#lx.\n", hr);
+    if (FAILED(hr))
+        goto done;
 
     /* Writing is allowed while storing, and the data will be stored by the next StoreAsync() call */
     hr = IDataWriter_WriteBytes(data_writer, 1, &byte_value);
@@ -1871,10 +1867,9 @@ static void test_DataWriter(void)
     todo_wine
     ok(hr == HRESULT_FROM_WIN32(ERROR_INVALID_OPERATION), "got hr %#lx.\n", hr);
 
+done:
     ref = IDataWriter_Release(data_writer);
     ok(ref == 0, "got ref %ld.\n", ref);
-
-done:
     ref = IRandomAccessStream_Release(in_memory_stream);
     ok(ref == 0, "got ref %ld.\n", ref);
     ref = IDataWriterFactory_Release(data_writer_factory);
-- 
2.55.0

